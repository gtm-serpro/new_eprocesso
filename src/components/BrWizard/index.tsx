import classNames from "classnames";
import React, {
  Children,
  ReactElement,
  useCallback,
  useEffect,
  useImperativeHandle,
  useRef,
  useState,
} from "react";
import useCommonProperties from "../../hooks/useCommonProperties";
import { useMtProps } from "../../hooks/useMtProps";
import { useSpreadProps } from "../../hooks/useSpreadProps";
import IMtProps from "../../types/IMtProps";
import WizardPanel from "./WizardPanel";

export interface BrWizardProps
  extends Omit<React.HTMLAttributes<HTMLDivElement>, "onChange">,
    IMtProps {
  /** Altura do Wizard. */
  height?: string;

  /** Se é do tipo vertical. Do contrário, é horizontal. */
  vertical?: boolean;

  /** Se mostra ou não o botão de cancelar */
  showCancelButton?: boolean;

  /** Texto do botão de cancelar */
  cancelButtonText?: string;

  /** Texto do botão para voltar ao passo anterior */
  prevButtonText?: string;

  /** Texto do botão de ir ao próximo passo. */
  nextButtonText?: string;

  /** Texto do botão de concluir. */
  concludeButtonText?: string;

  /** Número da aba inicial */
  initialStep?: number;

  /** Evento a ser disparado ao clicar no botão "Concluir" */
  onConclude?: (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;

  /** Evento a ser disparado ao clicar no botão "Cancelar" */
  onCancel?: (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;

  /** Evento a ser disparado ao se trocar de step */
  onChange?: (currentStep: number) => void;
}

const BrWizard = React.forwardRef<HTMLDivElement, BrWizardProps>(
  (
    {
      className,
      children,
      vertical = false,
      height = "400px",
      showCancelButton = false,
      cancelButtonText = "Cancelar",
      prevButtonText = "Voltar",
      nextButtonText = "Avançar",
      concludeButtonText = "Concluir",
      initialStep = 1,
      onConclude,
      onCancel,
      onChange,
      ...props
    },
    ref
  ) => {
    const mtProps = useMtProps(props);
    const spreadProps = useSpreadProps(props);
    const refWrapper = useRef(null);
    const refDiv = useRef(null);
    const refWizardProgress = useRef(null);
    const refWizardForm = useRef<HTMLDivElement>(null);
    const refConcludeButton = useRef(null);

    useCommonProperties<HTMLDivElement>(ref, refWrapper);

    const [currentStep, setCurrentStep] = useState<number>(initialStep || 1);
    const previousStep = useRef(initialStep || 1);

    useEffect(() => {
      previousStep.current = currentStep;
    }, [currentStep]);

    const handleClickStepButton = useCallback(
      (newStep: number) => {
        setCurrentStep(newStep);
        onChange?.(newStep);
      },
      [onChange]
    );

    const handleClickNextStep = useCallback(() => {
      setCurrentStep(previousStep.current + 1);
      onChange?.(previousStep.current + 1);
    }, [onChange]);

    const handleClickPrevStep = useCallback(() => {
      setCurrentStep(previousStep.current - 1);
      onChange?.(previousStep.current - 1);
    }, [onChange]);

    const nextStepComponent = Children.toArray(children)?.[currentStep] as ReactElement<any>;
    const isNextStepDisabled = !!nextStepComponent?.props.disabled;
    const prevStepComponent =
      currentStep > 1 ? (Children.toArray(children)?.[currentStep - 2] as ReactElement<any>) : null;
    const isPrevStepDisabled = !!prevStepComponent?.props.disabled;

    useImperativeHandle<HTMLDivElement, any>(ref, () => ({
      get div() {
        return refDiv.current;
      },
      get wrapper() {
        return refWrapper.current;
      },
      get wizardProgress() {
        return refWizardProgress.current;
      },
      get wizardForm() {
        return refWizardForm.current;
      },
      get concludeButton() {
        return refConcludeButton.current;
      },
    }));

    return (
      <div ref={refDiv} style={{ height: height }}>
        <div
          ref={refWrapper}
          className={classNames("br-wizard", className, ...mtProps)}
          {...spreadProps}
          {...(vertical && { vertical: "vertical" })}
          collapsed="collapsed"
          step={String(initialStep)}
        >
          <div ref={refWizardProgress} className="wizard-progress">
            {Children.map(children, (element: any, index) => (
              <button
                className="wizard-progress-btn"
                type="button"
                disabled={element.props.disabled}
                title={element.props.title}
                key={element.props.title}
                {...{ step: index + 1 }}
                onClick={() => handleClickStepButton(index + 1)}
                {...(currentStep === index + 1 && { active: "active" })}
              >
                <span className="info">{element.props.title}</span>
              </button>
            ))}
          </div>
          <div className="wizard-form" ref={refWizardForm}>
            {Children.map(children, (element, index) => (
              <div
                key={index}
                className="wizard-panel"
                {...(currentStep === index + 1 && { active: "active" })}
              >
                {element}
                <div className="wizard-panel-btn">
                  {showCancelButton && (
                    <button
                      className="br-button wizard-btn-canc"
                      type="button"
                      onClick={(event) => {
                        if (onCancel) onCancel(event);
                      }}
                    >
                      {cancelButtonText}
                    </button>
                  )}
                  {index < Children.count(children) - 1 && (
                    <button
                      onClick={handleClickNextStep}
                      disabled={isNextStepDisabled}
                      className="br-button primary wizard-btn-next"
                      type="button"
                    >
                      {nextButtonText}
                    </button>
                  )}
                  {index === Children.count(children) - 1 && (
                    <button
                      className="br-button primary wizard-btn"
                      type="button"
                      ref={refConcludeButton}
                      disabled={isNextStepDisabled}
                      onClick={(event) => {
                        if (onConclude) onConclude(event);
                      }}
                    >
                      {concludeButtonText}
                    </button>
                  )}
                  {index > 0 && !isPrevStepDisabled && (
                    <button
                      onClick={handleClickPrevStep}
                      className="br-button secondary wizard-btn-prev"
                      type="button"
                    >
                      {prevButtonText}
                    </button>
                  )}
                </div>
              </div>
            ))}
          </div>
        </div>
      </div>
    );
  }
);

export default Object.assign(BrWizard, {
  Panel: WizardPanel,
});
