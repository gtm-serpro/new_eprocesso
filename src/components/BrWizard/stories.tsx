import { Meta, Story } from "@storybook/react";

import BrWizard, { BrWizardProps } from ".";
import BrWizardPanel from "./WizardPanel/WizardPanel";

export default {
  title: "Navegação/Wizard",
  component: BrWizard,
} as Meta;

export const Simples: Story<BrWizardProps> = (args) => (
  <BrWizard {...args} onConclude={(event) => alert("Cliquei em Concluir")}>
    <BrWizardPanel title="Dados Pessoais">
      Conteúdo aqui - Lorem Ipsum is simply dummy text of the printing and typesetting industry.
      Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown
      printer took a galley of type and scrambled it to make a type specimen book. It has survived
      not only five centuries, but also the leap into electronic typesetting, remaining essentially
      unchanged.
    </BrWizardPanel>
    <BrWizardPanel title="Validar Dados">
      Conteúdo aqui - Lorem Ipsum is simply dummy text of the printing and typesetting industry.
      Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown
      printer took a galley of type and scrambled it to make a type specimen book. It has survived
      not only five centuries, but also the leap into electronic typesetting, remaining essentially
      unchanged.
    </BrWizardPanel>
    <BrWizardPanel title="Habilitar Cadastro">
      Conteúdo aqui - Lorem Ipsum is simply dummy text of the printing and typesetting industry.
      Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown
      printer took a galley of type and scrambled it to make a type specimen book. It has survived
      not only five centuries, but also the leap into electronic typesetting, remaining essentially
      unchanged.
    </BrWizardPanel>
    <BrWizardPanel title="Cadastrar Senha">
      Conteúdo aqui - Lorem Ipsum is simply dummy text of the printing and typesetting industry.
      Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown
      printer took a galley of type and scrambled it to make a type specimen book. It has survived
      not only five centuries, but also the leap into electronic typesetting, remaining essentially
      unchanged.
    </BrWizardPanel>
    <BrWizardPanel title="Finalizar">
      Conteúdo aqui - Lorem Ipsum is simply dummy text of the printing and typesetting industry.
      Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown
      printer took a galley of type and scrambled it to make a type specimen book. It has survived
      not only five centuries, but also the leap into electronic typesetting, remaining essentially
      unchanged.
    </BrWizardPanel>
  </BrWizard>
);
Simples.args = {
  height: "400px",
};
