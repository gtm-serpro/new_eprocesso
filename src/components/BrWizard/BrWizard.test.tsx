import { render, screen } from "@testing-library/react";

import BrWizard from ".";
import BrWizardPanel from "./WizardPanel/WizardPanel";

describe("<BrWizard />", () => {
  it("deveria renderizar o wizard", () => {
    const { container } = render(
      <BrWizard height="400px" onConclude={(event) => alert("Cliquei em Concluir")}>
        <BrWizardPanel title="Dados Pessoais">
          Conteúdo aqui - Lorem Ipsum is simply dummy text of the printing and typesetting industry.
          Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an
          unknown printer took a galley of type and scrambled it to make a type specimen book. It
          has survived not only five centuries, but also the leap into electronic typesetting,
          remaining essentially unchanged.
        </BrWizardPanel>
        <BrWizardPanel title="Validar Dados">
          Conteúdo aqui - Lorem Ipsum is simply dummy text of the printing and typesetting industry.
          Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an
          unknown printer took a galley of type and scrambled it to make a type specimen book. It
          has survived not only five centuries, but also the leap into electronic typesetting,
          remaining essentially unchanged.
        </BrWizardPanel>
        <BrWizardPanel title="Habilitar Cadastro">
          Conteúdo aqui - Lorem Ipsum is simply dummy text of the printing and typesetting industry.
          Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an
          unknown printer took a galley of type and scrambled it to make a type specimen book. It
          has survived not only five centuries, but also the leap into electronic typesetting,
          remaining essentially unchanged.
        </BrWizardPanel>
        <BrWizardPanel title="Cadastrar Senha">
          Conteúdo aqui - Lorem Ipsum is simply dummy text of the printing and typesetting industry.
          Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an
          unknown printer took a galley of type and scrambled it to make a type specimen book. It
          has survived not only five centuries, but also the leap into electronic typesetting,
          remaining essentially unchanged.
        </BrWizardPanel>
        <BrWizardPanel title="Finalizar">
          Conteúdo aqui - Lorem Ipsum is simply dummy text of the printing and typesetting industry.
          Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an
          unknown printer took a galley of type and scrambled it to make a type specimen book. It
          has survived not only five centuries, but also the leap into electronic typesetting,
          remaining essentially unchanged.
        </BrWizardPanel>
      </BrWizard>
    );

    expect(container.firstChild?.firstChild).toHaveClass("br-wizard");
    expect(container.firstChild).toMatchSnapshot();
  });
});
