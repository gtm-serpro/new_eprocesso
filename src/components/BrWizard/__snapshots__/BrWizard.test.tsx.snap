// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BrWizard /> deveria renderizar o wizard 1`] = `
<div
  style="height: 400px;"
>
  <div
    class="br-wizard"
    collapsed="collapsed"
    step="1"
  >
    <div
      class="wizard-progress"
    >
      <button
        active="active"
        class="wizard-progress-btn"
        step="1"
        title="Dados Pessoais"
        type="button"
      >
        <span
          class="info"
        >
          Dados Pessoais
        </span>
      </button>
      <button
        class="wizard-progress-btn"
        step="2"
        title="Validar Dados"
        type="button"
      >
        <span
          class="info"
        >
          Validar Dados
        </span>
      </button>
      <button
        class="wizard-progress-btn"
        step="3"
        title="Habilitar Cadastro"
        type="button"
      >
        <span
          class="info"
        >
          Habilitar Cadastro
        </span>
      </button>
      <button
        class="wizard-progress-btn"
        step="4"
        title="Cadastrar Senha"
        type="button"
      >
        <span
          class="info"
        >
          Cadastrar Senha
        </span>
      </button>
      <button
        class="wizard-progress-btn"
        step="5"
        title="Finalizar"
        type="button"
      >
        <span
          class="info"
        >
          Finalizar
        </span>
      </button>
    </div>
    <div
      class="wizard-form"
    >
      <div
        active="active"
        class="wizard-panel"
      >
        <div
          class="wizard-panel-content"
        >
          <div
            class="h3"
          >
            Dados Pessoais
          </div>
          <div
            class="text"
            tabindex="0"
          >
            Conteúdo aqui - Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged.
          </div>
        </div>
        <div
          class="wizard-panel-btn"
        >
          <button
            class="br-button primary wizard-btn-next"
            type="button"
          >
            Avançar
          </button>
        </div>
      </div>
      <div
        class="wizard-panel"
      >
        <div
          class="wizard-panel-content"
        >
          <div
            class="h3"
          >
            Validar Dados
          </div>
          <div
            class="text"
            tabindex="0"
          >
            Conteúdo aqui - Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged.
          </div>
        </div>
        <div
          class="wizard-panel-btn"
        >
          <button
            class="br-button primary wizard-btn-next"
            type="button"
          >
            Avançar
          </button>
          <button
            class="br-button secondary wizard-btn-prev"
            type="button"
          >
            Voltar
          </button>
        </div>
      </div>
      <div
        class="wizard-panel"
      >
        <div
          class="wizard-panel-content"
        >
          <div
            class="h3"
          >
            Habilitar Cadastro
          </div>
          <div
            class="text"
            tabindex="0"
          >
            Conteúdo aqui - Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged.
          </div>
        </div>
        <div
          class="wizard-panel-btn"
        >
          <button
            class="br-button primary wizard-btn-next"
            type="button"
          >
            Avançar
          </button>
          <button
            class="br-button secondary wizard-btn-prev"
            type="button"
          >
            Voltar
          </button>
        </div>
      </div>
      <div
        class="wizard-panel"
      >
        <div
          class="wizard-panel-content"
        >
          <div
            class="h3"
          >
            Cadastrar Senha
          </div>
          <div
            class="text"
            tabindex="0"
          >
            Conteúdo aqui - Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged.
          </div>
        </div>
        <div
          class="wizard-panel-btn"
        >
          <button
            class="br-button primary wizard-btn-next"
            type="button"
          >
            Avançar
          </button>
          <button
            class="br-button secondary wizard-btn-prev"
            type="button"
          >
            Voltar
          </button>
        </div>
      </div>
      <div
        class="wizard-panel"
      >
        <div
          class="wizard-panel-content"
        >
          <div
            class="h3"
          >
            Finalizar
          </div>
          <div
            class="text"
            tabindex="0"
          >
            Conteúdo aqui - Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged.
          </div>
        </div>
        <div
          class="wizard-panel-btn"
        >
          <button
            class="br-button primary wizard-btn"
            type="button"
          >
            Concluir
          </button>
          <button
            class="br-button secondary wizard-btn-prev"
            type="button"
          >
            Voltar
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
